src/main/java/io/nats/client/impl/NatsMessage.java

```diff
@@ -26,103 +26,75 @@
-
-    static final byte[] digits = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9'};
-
-    static int copy(byte[] dest, int pos, String toCopy) {
-        for (int i=0, max=toCopy.length(); i<max ;i++) {
-            dest[pos] = (byte) toCopy.charAt(i);
-            pos++;
-        }
+    private Integer protocolLength = null;
 
-        return pos;
-    }
-
-    private static String PUB_SPACE = NatsConnection.OP_PUB + " ";
-    private static String SPACE = " ";
+    NatsMessage next; // for linked list
 
     // Create a message to publish
-    NatsMessage(String subject, String replyTo, byte[] data, boolean utf8mode) {
+    NatsMessage(String subject, String replyTo, ByteBuffer data, boolean utf8mode) {
         this.subject = subject;
         this.replyTo = replyTo;
         this.data = data;
-        
-        if (utf8mode) {
-            int subjectSize = subject.length() * 2;
-            int replySize = (replyTo != null) ? replyTo.length() * 2 : 0;
-            StringBuilder protocolStringBuilder = new StringBuilder(4 + subjectSize + 1 + replySize + 1);
-            protocolStringBuilder.append(PUB_SPACE);
-            protocolStringBuilder.append(subject);
-            protocolStringBuilder.append(SPACE);
-    
-            if (replyTo != null) {
-                protocolStringBuilder.append(replyTo);
-                protocolStringBuilder.append(SPACE);
-            }
-    
-            protocolStringBuilder.append(String.valueOf(data.length));
+        Charset charset;
 
-            this.protocolBytes = protocolStringBuilder.toString().getBytes(StandardCharsets.UTF_8);
-        } else {
-            // Convert the length to bytes
-            byte[] lengthBytes = new byte[12];
-            int idx = lengthBytes.length;
-            int size = (data != null) ? data.length : 0;
-
-            if (size > 0) {
-                for (int i = size; i > 0; i /= 10) {
-                    idx--;
-                    lengthBytes[idx] = digits[i % 10];
-                }
+        // Calculate the length in bytes
+        int size = (data != null) ? data.limit() : 0;
+        int len = 4;
+        len += fastIntLength(size);
+        if (replyTo != null) {
+            if (utf8mode) {
+                len += fastUtf8Length(replyTo) + 1;
             } else {
-                idx--;
-                lengthBytes[idx] = digits[0];
-            }
-
-            // Build the array
-            int len = 4 + subject.length() + 1 + (lengthBytes.length - idx);
-
-            if (replyTo != null) {
                 len += replyTo.length() + 1;
             }
+        }
+        if (utf8mode) {
+            len += fastUtf8Length(subject) + 1;
+            charset = StandardCharsets.UTF_8;
+        } else {
+            len += subject.length() + 1;
+            charset = StandardCharsets.US_ASCII;
+        }
+        this.protocolBytes = ByteBuffer.allocate(len);
+        protocolBytes.put((byte)'P').put((byte)'U').put((byte)'B').put((byte)' ');

```

Comment:
Would it be faster to use a backing array, initialized with `PUB ` vs put APIs?  Not sure which is faster in java...

Reply:
I recall this was faster, and it will have better interoperability with the next round of changes which use `ByteBuffer`'s in more places, `ByteBuffer` should be optimized for these types of operations.